dropshot/src/server.rs

```diff
@@ -543,18 +545,83 @@ impl<C: ServerContext> Service<&TlsConn> for ServerConnectionHandler<C> {
     }
 }
 
-/**
- * A running Dropshot HTTP server.
- */
-pub struct HttpServer<C: ServerContext> {
+/// Marker trait used to control whether or not an HTTP server can be closed.
+pub trait Closer: Send + Sync + Unpin + 'static {}
```

Comment:
I had a couple thoughts here:

1. Do we actually need the `Closer` / `Joiner` traits? There's no way for a caller to provide new implementors of these types, and we only ever have 2 concrete marker types for each one.
2. Could we move the `HttpServerCloser` / `HttpServerJoiner` into `Closeable` / `Joinable`, removing the need to hold `Option<_>`s at the `HttpServer` level?

I had to implement this to convince myself this would work; I think it does: https://github.com/oxidecomputer/dropshot/compare/cleanup...cleanup-tweak. A nice side effect is that our `Future`/`FusedFuture` impls can call down to the `HttpServerJoiner` implementations instead of copying them.

Reply:
I like that suggestion a lot; pulled it into https://github.com/oxidecomputer/dropshot/pull/537/commits/a06a78a67828d9368dd6c45d6fbf94ba7fc06866

Reply:
I went one step further and flattened the structures even more in b2ce08def4659b7c6694f0a87b04f5f22b5da858 , which IMO also simplifies the readability -- no need for tuple structs